### PR TITLE
Add Rayon-based parallel iterator over GraphMap nodes

### PR DESCRIPTION
**Pull request by daniel-levin**
See #572 for some context. In short, I have a use case for parallel iteration over the nodes in the underlying graph. I don't want every user of this crate to have to include Rayon, so I added it as a conditional dependency, as it is in indexmap. I wanted to create this PR early, to get an idea of the desirability of such a change - if this is something the petgraph team wants, I'll continue adding more.

The purpose of most of the code I've written so far is to expose a parallel iterator API that's doesn't re-export any types from indexmap and doesn't use any crate-private types.

I've also added a microbenchmark replicating my real-world workload.

I'm running the benchmarks on my laptop's i7-7500U, and even though [benchmarking on a laptop is a sin](https://lemire.me/blog/my-sayings/), the difference between serial and parallel is striking:

```
running 2 tests
test graphmap_parallel_bench ... bench:     707,148 ns/iter (+/- 325,762)
test graphmap_serial_bench   ... bench:   1,635,905 ns/iter (+/- 379,030)
```

Edit: ran the benchmarks on a m6g.2xlarge on EC2:
```
running 2 tests
test graphmap_parallel_bench ... bench:     422,702 ns/iter (+/- 4,520)
test graphmap_serial_bench   ... bench:   2,998,678 ns/iter (+/- 7,824)
```

Edit: I've edited the benchmarks to do something 'real', which still produces considerable speedups:

```
running 2 tests
test graphmap_parallel_bench ... bench:  35,571,906 ns/iter (+/- 2,379,332)
test graphmap_serial_bench   ... bench: 285,738,589 ns/iter (+/- 23,645,030)
```

### CODE REVIEW

File: src/rayon/mod.rs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+/// Adds parallel iterators where possible, i.e. where nodes are [Send] + [Sync].
+pub trait NodesParIter<'a, N>
+where
+    N: Send + Sync,
+{
+    /// Returns a parallel iterator over this graph's nodes.
+    fn par_nodes(&'a self) -> ParNodes<'a, N>;
+}
+
+impl<'a, N, E, Ty> NodesParIter<'a, N> for GraphMap<N, E, Ty>
```
Comment (cuviper) [inline src/rayon/mod.rs:36]: Why use an extension trait for this? An inherent method on `GraphMap` should work fine, even defined from here in this module.

Comment (daniel-levin) [reply src/rayon/mod.rs:36]: I didn't know Rust lets you do that 😅

**Comment by daniel-levin** [comment]
Good callout @cuviper ! I was not aware that you could define inherent methods in different modules, but now that I know that, I agree it's the right thing to do here. I've updated the code to do this.

File: Cargo.toml [hunk 2]
```diff
@@ -38,10 +38,11 @@ debug = true
 
 [dependencies]
 fixedbitset = { version = "0.4.0", default-features = false }
-indexmap = "2.0"
+indexmap = { version = "2.0", default-features = true }
```
Comment (cuviper) [inline Cargo.toml:41]: What's the point of this? Defaults are on by default. :wink:

File: Cargo.toml [hunk 3]
```diff
@@ -51,9 +52,10 @@ odds = { version = "0.4.0" }
 rand = "0.5.5"
 
 [features]
+rayon = ["graphmap", "dep:rayon", "indexmap/rayon"]
```
Comment (cuviper) [inline Cargo.toml:55]: I don't think this should force `graphmap` on -- that happens to be the only useful place for rayon functionality _right now_, but there could be others later.

File: src/graphmap.rs [hunk 4]
```diff
@@ -59,7 +59,7 @@ pub type DiGraphMap<N, E> = GraphMap<N, E, Directed>;
 /// Depends on crate feature `graphmap` (default).
 #[derive(Clone)]
 pub struct GraphMap<N, E, Ty> {
-    nodes: IndexMap<N, Vec<(N, CompactDirection)>>,
+    pub(crate) nodes: IndexMap<N, Vec<(N, CompactDirection)>>,
```
Comment (cuviper) [inline src/graphmap.rs:62]: If these `pub(crate)` additions are not wanted, the `par_nodes` definition could be moved into this module or a submodule.

File: benches/graphmap.rs [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    gr
+}
+
+#[bench]
+fn graphmap_serial_bench(bench: &mut Bencher) {
+    let data = test_nodes();
+    let gr = test_graph(&data);
+    bench.iter(|| {
+        for n in gr.nodes() {
+            gr.edges_directed(n, Direction::Outgoing);
```
Comment (cuviper) [inline benches/graphmap.rs:53]: This seems like a strange workload, creating an iterator without consuming it...

**Review by cuviper — COMMENTED** [review]
I'm not a maintainer here, but since I was requested, I can add my 2 cents...

File: Cargo.toml (hunk 2 above)
Comment (daniel-levin) [reply Cargo.toml:41]: D'oh 😣 chalk it up to inexperience with Cargo

File: benches/graphmap.rs (hunk 5 above)
Comment (daniel-levin) [reply benches/graphmap.rs:53]: Yeah this is peculiar. I'm going to consume the iterator for the sake of clarity of the benchmark. It's kinda subtle. The construction of the inner iterator (despite it not being consumed) is still time consuming enough to bottleneck the outer iterator - which happens to be a minimal reproducible example of my use case - and enough to manifest in the profiling data.

File: Cargo.toml (hunk 3 above)
Comment (daniel-levin) [reply Cargo.toml:55]: Agreed.

**Comment by daniel-levin** [comment]
@bluss @ABorgna Kindly requesting that you take a look.

File: Cargo.toml (hunk 2 above)
Comment (ABorgna) [reply Cargo.toml:41]: Agreed with the suggestion,
```suggestion
indexmap = "2.0"
```

File: src/graphmap.rs (hunk 4 above)
Comment (ABorgna) [reply src/graphmap.rs:62]: Yes, I say move `par_nodes` to here instead.

**Review by ABorgna — COMMENTED** [review]
Nice! I don't have more to add to @cuviper's review; just move the code to `graphmap.rs`.

File: Cargo.toml (hunk 2 above)
Comment (daniel-levin) [reply Cargo.toml]: Fixed this. At the time I misunderstood how feature flags worked.

File: src/graphmap.rs (hunk 4 above)
Comment (daniel-levin) [reply src/graphmap.rs]: Done. I've removed the `rayon` module. Everything has been put into `graphmap.rs`, which on reflection, is quite a bit nicer than what I first proposed.

**Comment by daniel-levin** [comment]
> Nice! I don't have more to add to @cuviper's review; just move the code to `graphmap.rs`.

Done! Thanks for the comments.

**Pull request merged**